include: "rules/00_common.smk"

##### Target rules #####

rule all:
    input:
################
## 00_common.smk
################
## create_f2_samples_file
#        config["F2_samples"],
################
## 01_ref.smk
################
## get_genome
#        config["ref_prefix"] + ".fasta",
## genome_faidx
#        config["ref_prefix"] + ".fasta.fai",
## genome_dict
#        config["ref_prefix"] + ".dict",
################
## 02_F0_mapping.smk
################
## map_reads_F0
#        expand(os.path.join(config["working_dir"], "sams/F0/mapped/{F0_sample}-{unit}.sam"),
#                zip,
#                F0_sample = F0_samples['sample'],
#                unit = F0_samples['unit']
#        ),
## replace_rg_F0
#        expand(os.path.join(config["working_dir"], "sams/F0/grouped/{F0_sample}-{unit}.sam"),
#                zip,
#                F0_sample = F0_samples['sample'],
#                unit = F0_samples['unit']
#        ),
## sort_sam_F0
#        expand(os.path.join(config["working_dir"], "bams/F0/sorted/{F0_sample}-{unit}.bam"),
#                zip,
#                F0_sample = F0_samples['sample'],
#                unit = F0_samples['unit']
#        ),
## mark_duplicates_F0
#        expand(os.path.join(config["working_dir"], "bams/F0/marked/{F0_sample}-{unit}.bam"),
#                zip,
#                F0_sample = F0_samples['sample'],
#                unit = F0_samples['unit']                
#        ),
## merge_bams_F0
#        expand(os.path.join(config["working_dir"], "bams/F0/merged/{F0_sample}.bam"),
#                F0_sample = config["F0_lines"]
#        ),
## samtools_index_F0
#        expand(os.path.join(config["working_dir"], "bams/F0/merged/{F0_sample}.bam.bai"),
#                F0_sample = config["F0_lines"]
#        ),
################
## 03_F0_calling.smk
################
## haplotype_caller
#        expand(os.path.join(config["working_dir"], "vcfs/F0/gvcfs/{F0_sample}/{contig}.g.vcf"),
#                F0_sample = config["F0_lines"],
#                contig = get_contigs(),
#        ),
## combine_calls
#        expand(os.path.join(config["working_dir"], "vcfs/F0/combined/all.{contig}.g.vcf.gz"),
#                contig = get_contigs(),
#        ),
##Â genotype_variants
#        expand(os.path.join(config["working_dir"], "vcfs/F0/genotyped/all.{contig}.vcf.gz"),
#                contig = get_contigs(),
#        ),
## merge_variants
#        os.path.join(config["data_store_dir"], "vcfs/F0/final/all.vcf.gz"),
## merge_variants_F0_and_F2
#        os.path.join(config["data_store_dir"], "vcfs/F0_and_F2/final/all.vcf.gz"),
###############
# 04_F2_mapping.smk
###############
## copy_f2_seq_data
#        expand(os.path.join(config["working_dir"], "fastqs/F2/{F2_sample}_{pair}.txt.gz"),
#                F2_sample = F2_samples['SAMPLE'],
#                pair = config["pairs"]
#        ),
## bwa_mem2_mem
#        expand(os.path.join(config["working_dir"], "sams/F2/bwamem2/mapped/{F2_sample}.sam"),
#                F2_sample = F2_samples['SAMPLE']
#        ),
## sort_sam_f2
#        expand(os.path.join(config["working_dir"], "bams/F2/bwamem2/sorted/{F2_sample}.bam"),
#                F2_sample = F2_samples['SAMPLE']
#        ),
## mark_duplicates_f2
#        expand(os.path.join(config["working_dir"], "bams/F2/bwamem2/marked/{F2_sample}.bam"),
#                F2_sample = F2_samples['SAMPLE']
#        ),
## samtools_index_f2
#        expand(os.path.join(config["working_dir"], "bams/F2/bwamem2/marked/{F2_sample}.bam.bai"),
#                F2_sample = F2_samples['SAMPLE']
#        ),
################
## 05_F2_recomb_blocks.smk
################
## get_homozygous_sites
#        expand(os.path.join(config["working_dir"], "data/sites_files/F0_Cab_Kaga/homo_all/{F0_sample}.csv"),
#                F0_sample = config["F0_lines"]
#        ),
## exclude_repeat_sites_F2
#        os.path.join(config["working_dir"], "data/sites_files/F0_Cab_Kaga/homo_divergent/no_repeats.txt"),
## get_divergent_sites
#        os.path.join(config["working_dir"], "data/sites_files/F0_Cab_Kaga/homo_divergent/all.txt"),
## bam_readcount_F2
#        expand(os.path.join(config["working_dir"], "dp4s/batch_01/bwamem2/{F2_sample}.dp4.txt"),
#                F2_sample = F2_samples['SAMPLE']
#        ),
## make_dp_AB_F2
#        expand(os.path.join(config["working_dir"], "dpABs/batch_01/bwamem2/{F2_sample}.txt"),
#                F2_sample = F2_samples['SAMPLE']
#        ),
## run_rc_block_F2
#        expand(os.path.join(config["data_store_dir"], "recombination_blocks/{date_of_processing}/F2_{bin_length}.txt"),
#                date_of_processing = config["date_of_processing"],
#                bin_length = config["bin_length"]
#        ),
################
## 06_F0_recomb_blocks.smk
################
## run_rc_block_F0
#        expand(os.path.join(config["data_store_dir"], "recombination_blocks/F0/{site_filter}/{bin_length}.txt"),
#                site_filter = config["site_filter"],
#                bin_length = config["bin_length_F0"]
#        ),
################
## 07_association_testing.smk
###############
## create_gwas_input
#        expand(os.path.join(config["data_store_dir"], "association_testing/{date_of_assoc_test}_true/inputs/{bin_length}.rds"),
#                date_of_assoc_test = config["date_of_assoc_test"],
#                bin_length = config["bin_length"]
#        ),
## test_gwls
#        expand(os.path.join(config["data_store_dir"], "association_testing/{date_of_assoc_test}_test/test_results/{bin_length}.rds"),
#                date_of_assoc_test = config["date_of_assoc_test"],
#                target_phenotype = config["target_phenotypes"],
#                bin_length = config["bin_length"]
#        ),
## run_gwls
#        expand(os.path.join(config["data_store_dir"], "association_testing/{date_of_assoc_test}_true/{site_filter}/results/{target_phenotype}/{bin_length}.rds"),
#                date_of_assoc_test = config["date_of_assoc_test"],
#                site_filter = config["site_filter"],
#                target_phenotype = config["target_phenotypes"],
#                bin_length = config["bin_length"]
#        ),
## run permutations
#        expand(os.path.join(config["data_store_dir"], "association_testing/{date_of_assoc_test}_permutations/{site_filter}/{target_phenotype}/{bin_length}/{permutation_seed}.rds"),
#                date_of_assoc_test = config["date_of_assoc_test"],
#                site_filter = config["site_filter"],
#                target_phenotype = config["target_phenotypes"],
#                bin_length = config["bin_length"],
#                permutation_seed = PERM_SEEDS      
#        ),
#
################
## 08_F2_calling.smk
###############
## merge_variants_F2
#        os.path.join(config["data_store_dir"], "vcfs/F2/final/all.vcf.gz"),
################
## 09_TIGER.smk
###############
# run_basecaller
        expand(os.path.join(config["working_dir"], "TIGER/01_base_called/{F2_sample}.txt"),
                F2_sample = F2_samples['SAMPLE']
        ),
# estimate_allele_freqs
        expand(os.path.join(config["working_dir"], "TIGER/02_af_estimated/{F2_sample}.txt"),
                F2_sample = F2_samples['SAMPLE']
        ),
# apply_beta_mixture_model
        expand(os.path.join(config["working_dir"], "TIGER/03_beta_modelled/{F2_sample}.txt"),
                F2_sample = F2_samples['SAMPLE']
        ),
# calculate_HMM_probs
        expand(os.path.join(config["working_dir"], "TIGER/05_hmm_probs/{F2_sample}.txt"),
                F2_sample = F2_samples['SAMPLE'][0]
        )

##### Modules #####

include: "rules/01_ref.smk"
include: "rules/02_F0_mapping.smk"
include: "rules/03_F0_calling.smk"
include: "rules/04_F2_mapping.smk"
include: "rules/05_F2_recomb_blocks.smk"
include: "rules/06_F0_recomb_blocks.smk"
include: "rules/07_assocation_testing.smk"
include: "rules/08_F2_calling.smk"
include: "rules/09_TIGER.smk"