<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 9 Association testing with only AU microscope | Somite period F2 analysis</title>
<meta name="author" content="Ian Brettell">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.2"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.10/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.5.1/tabs.js"></script><script src="libs/bs3compat-0.2.5.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="libs/plotly-binding-4.9.4.1/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.18/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><style>
    #content {
        font-size: 0.95rem;
    }
    </style>
<script src="https://cdn.jsdelivr.net/autocomplete.js/0/autocomplete.jquery.min.js"></script><script src="https://cdn.jsdelivr.net/npm/mark.js@8.11.1/dist/mark.min.js"></script><!-- CSS --><link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Somite period F2 analysis</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="alignment-of-f0-and-f2-generations.html"><span class="header-section-number">2</span> Alignment of F0 and F2 generations</a></li>
<li><a class="" href="homozygosity-of-f0-cab-and-kaga.html"><span class="header-section-number">3</span> Homozygosity of F0 Cab and Kaga</a></li>
<li><a class="" href="f2-recombination-blocks-all-homozygous-divergent-sites.html"><span class="header-section-number">4</span> F2 recombination blocks (all homozygous-divergent sites)</a></li>
<li><a class="" href="f2-recombination-blocks-filtered-sites.html"><span class="header-section-number">5</span> F2 recombination blocks (filtered sites)</a></li>
<li><a class="" href="phenotypes.html"><span class="header-section-number">6</span> Phenotypes</a></li>
<li><a class="" href="association-testing-with-simulated-phenotypes.html"><span class="header-section-number">7</span> Association testing with simulated phenotypes</a></li>
<li><a class="" href="gwas-results.html"><span class="header-section-number">8</span> GWAS results</a></li>
<li><a class="active" href="association-testing-with-only-au-microscope.html"><span class="header-section-number">9</span> Association testing with only AU microscope</a></li>
<li><a class="" href="association-testing-with-only-db-microscope.html"><span class="header-section-number">10</span> Association testing with only DB microscope</a></li>
<li><a class="" href="inverse-normalize-by-microscope.html"><span class="header-section-number">11</span> Inverse-normalize by microscope</a></li>
<li><a class="" href="annotations-of-gwas-hits.html"><span class="header-section-number">12</span> Annotations of GWAS hits</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/brettellebi/somites">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="association-testing-with-only-au-microscope" class="section level1" number="9">
<h1>
<span class="header-section-number">9</span> Association testing with only AU microscope<a class="anchor" aria-label="anchor" href="#association-testing-with-only-au-microscope"><i class="fas fa-link"></i></a>
</h1>
<div id="load-libraries-and-variables" class="section level2" number="9.1">
<h2>
<span class="header-section-number">9.1</span> Load libraries and variables<a class="anchor" aria-label="anchor" href="#load-libraries-and-variables"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GridLMM</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org">KRLS</a></span><span class="op">)</span>

<span class="va">GENO_FILE</span> <span class="op">=</span> <span class="st">"/nfs/research/birney/users/ian/somites/association_testing/20220214/all_sites/inputs/5000.rds"</span>
<span class="va">PHENO_FILE</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"data/20220214_phenotypes.xlsx"</span><span class="op">)</span> <span class="co"># True phenotypes</span>
<span class="va">GWLS_SOURCE_FILE</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"workflow/scripts/run_gwls_source.R"</span><span class="op">)</span>
<span class="va">MANHAT_SOURCE_FILE</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"workflow/scripts/get_manhattan_source.R"</span><span class="op">)</span>
<span class="va">BIN_LENGTH</span> <span class="op">=</span> <span class="fl">5000</span>
<span class="va">TARGET_PHENO</span> <span class="op">=</span> <span class="st">"intercept"</span>
<span class="va">MICROSCOPE</span> <span class="op">=</span> <span class="st">"AU"</span>
<span class="va">PLOT_DIR</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"book/plots/20220214/microscope_test"</span><span class="op">)</span>
<span class="va">ALPHA</span> <span class="op">=</span> <span class="fl">0.05</span>

<span class="co"># Get GWAS functions</span>

<span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="va">GWLS_SOURCE_FILE</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="va">MANHAT_SOURCE_FILE</span><span class="op">)</span></code></pre></div>
</div>
<div id="load-genotypes-and-positions" class="section level2" number="9.2">
<h2>
<span class="header-section-number">9.2</span> Load genotypes and positions<a class="anchor" aria-label="anchor" href="#load-genotypes-and-positions"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">in_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">GENO_FILE</span><span class="op">)</span></code></pre></div>
</div>
<div id="read-in-phenotypes" class="section level2" number="9.3">
<h2>
<span class="header-section-number">9.3</span> Read in phenotypes<a class="anchor" aria-label="anchor" href="#read-in-phenotypes"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R">

<span class="co">## Read in file and wrangle</span>
<span class="va">phenos</span> <span class="op">=</span> <span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_xlsx</a></span><span class="op">(</span><span class="va">PHENO_FILE</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># adjust sample names</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>SAMPLE <span class="op">=</span> <span class="va">fish</span> <span class="op">%&gt;%</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="st">"KC"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># select key columns</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">SAMPLE</span>, <span class="fu">all_of</span><span class="op">(</span><span class="va">TARGET_PHENO</span><span class="op">)</span>, <span class="va">Microscope</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># ensure that the phenotype column is numeric</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">TARGET_PHENO</span><span class="op">)</span>,
                                <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
    <span class="co"># THIS IS THE (big) CHANGE: filter for microscope</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Microscope</span> <span class="op">==</span> <span class="va">MICROSCOPE</span><span class="op">)</span></code></pre></div>
</div>
<div id="filter-for-samples-with-both-genos-and-phenos" class="section level2" number="9.4">
<h2>
<span class="header-section-number">9.4</span> Filter for samples with both genos and phenos<a class="anchor" aria-label="anchor" href="#filter-for-samples-with-both-genos-and-phenos"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Filter and order phenotypes</span>
<span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span>
    <span class="co"># filter phenotypes for those with genotypes</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">SAMPLE</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># join to `sample_order` to ensure phenotypes are in the correct order   </span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>SAMPLE <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                     <span class="va">.</span>,
                     by <span class="op">=</span> <span class="st">"SAMPLE"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># remove NAs (created by the samples that have genotypes but not phenotypes)</span>
    <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html">drop_na</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># the GridLMM code doesn't work with tibbles</span>
    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## Filter genotypes for those that have phenotypes</span>
<span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span> <span class="op">%&gt;%</span> 
                   <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Filter sample_order for those that have phenotypes</span>
<span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span>

<span class="co">## Get number of samples</span>
<span class="va">N_SAMPLES</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-gwas" class="section level2" number="9.5">
<h2>
<span class="header-section-number">9.5</span> Run GWAS<a class="anchor" aria-label="anchor" href="#run-gwas"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run GWAS</span>

<span class="va">out</span> <span class="op">=</span> <span class="fu">run_gwas</span><span class="op">(</span>d <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span>,
               m <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"positions"</span><span class="op">]</span><span class="op">]</span>,
               p <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span>,
               invers_norm <span class="op">=</span> <span class="cn">T</span>
              <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">out</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/gwas_results/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb112"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/gwas_results/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-permutations" class="section level2" number="9.6">
<h2>
<span class="header-section-number">9.6</span> Run permutations<a class="anchor" aria-label="anchor" href="#run-permutations"><i class="fas fa-link"></i></a>
</h2>
<div id="permute-phenos" class="section level3" number="9.6.1">
<h3>
<span class="header-section-number">9.6.1</span> Permute phenos<a class="anchor" aria-label="anchor" href="#permute-phenos"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seeds</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>

<span class="va">counter</span> <span class="op">=</span> <span class="fl">0</span>
<span class="va">perm_phenos</span> <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">seeds</span>, <span class="kw">function</span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span><span class="op">{</span>
  <span class="va">counter</span> <span class="op">&lt;&lt;-</span> <span class="va">counter</span> <span class="op">+</span> <span class="fl">1</span>
  <span class="co"># set seed</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">seeds</span><span class="op">[</span><span class="va">counter</span><span class="op">]</span><span class="op">)</span>
  <span class="co"># get original phenotypes</span>
  <span class="va">orig_phenos</span> <span class="op">=</span> <span class="va">phenos</span>
  <span class="co"># randomise</span>
  <span class="va">phenos</span> <span class="op">=</span> <span class="va">orig_phenos</span>
      <span class="co"># randomise phenotype</span>
      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">TARGET_PHENO</span><span class="op">)</span>,
                                  <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-gwls" class="section level3" number="9.6.2">
<h3>
<span class="header-section-number">9.6.2</span> Run GWLS<a class="anchor" aria-label="anchor" href="#run-gwls"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb114"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm_out</span> <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">perm_phenos</span>, <span class="kw">function</span><span class="op">(</span><span class="va">PERM_PHENO</span><span class="op">)</span><span class="op">{</span>
  <span class="co">## Get phenotypes</span>
  <span class="va">phenos</span> <span class="op">=</span> <span class="va">PERM_PHENO</span>
  
  <span class="co">## Filter and order phenotypes</span>
  <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span>
      <span class="co"># filter phenotypes for those with genotypes</span>
      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">SAMPLE</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="co"># join to `sample_order` to ensure phenotypes are in the correct order   </span>
      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>SAMPLE <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                       <span class="va">.</span>,
                       by <span class="op">=</span> <span class="st">"SAMPLE"</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="co"># remove NAs (created by the samples that have genotypes but not phenotypes)</span>
      <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html">drop_na</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="co"># the GridLMM code doesn't work with tibbles</span>
      <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
  
  <span class="co">## Filter genotypes for those that have phenotypes</span>
  <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span>
      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span><span class="op">)</span>
  
  <span class="co">## Filter sample_order for those that have phenotypes</span>
  <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span>
              
  <span class="co"># Run GWAS</span>
  
  <span class="va">out</span> <span class="op">=</span> <span class="fu">run_gwas</span><span class="op">(</span>d <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span>,
                 m <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"positions"</span><span class="op">]</span><span class="op">]</span>,
                 p <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span>,
                 invers_norm <span class="op">=</span> <span class="cn">T</span>
                <span class="op">)</span>
  
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">perm_out</span><span class="op">)</span> <span class="op">=</span> <span class="va">seeds</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">perm_out</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/perms/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm_out</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/perms/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="get-minimum" class="section level3" number="9.6.3">
<h3>
<span class="header-section-number">9.6.3</span> Get minimum<a class="anchor" aria-label="anchor" href="#get-minimum"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb116"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm_df</span> <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span><span class="op">(</span><span class="va">perm_out</span>, <span class="kw">function</span><span class="op">(</span><span class="va">PERM</span><span class="op">)</span><span class="op">{</span>
  <span class="va">OUT</span> <span class="op">=</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>MIN_P <span class="op">=</span> <span class="va">PERM</span><span class="op">$</span><span class="va">results</span><span class="op">$</span><span class="va">p_value_REML</span> <span class="op">%&gt;%</span>
                         <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">.</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
  <span class="op">)</span>
<span class="op">}</span>, .id <span class="op">=</span> <span class="st">"SEED"</span><span class="op">)</span>

<span class="co"># Get minimum</span>
<span class="va">SIG_LEVEL</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">perm_df</span><span class="op">$</span><span class="va">MIN_P</span><span class="op">)</span>

<span class="co"># Get bonferroni correction</span>
<span class="va">SIG_BONF</span> <span class="op">=</span> <span class="va">ALPHA</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span><span class="op">(</span><span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="generate-manhattan-plot" class="section level2" number="9.7">
<h2>
<span class="header-section-number">9.7</span> Generate Manhattan plot<a class="anchor" aria-label="anchor" href="#generate-manhattan-plot"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out_clean</span> <span class="op">=</span> <span class="fu">clean_gwas_res</span><span class="op">(</span><span class="va">out</span>,
                           bin_length <span class="op">=</span> <span class="va">BIN_LENGTH</span>,
                           chr_lens <span class="op">=</span> <span class="va">med_chr_lens</span><span class="op">)</span>

<span class="co"># Plot</span>
<span class="va">out_plot</span> <span class="op">=</span> <span class="fu">plot_man</span><span class="op">(</span><span class="va">out_clean</span>,
                    site_filter <span class="op">=</span> <span class="st">"all_sites"</span>,
                    phenotype <span class="op">=</span> <span class="va">TARGET_PHENO</span>,
                    bin_length <span class="op">=</span> <span class="va">BIN_LENGTH</span>, 
                    gwas_pal <span class="op">=</span> <span class="va">intercept_pal</span>,
                    med_chr_lens <span class="op">=</span> <span class="va">med_chr_lens</span>,
                    sig_level <span class="op">=</span> <span class="va">SIG_LEVEL</span>,
                    bonferroni <span class="op">=</span> <span class="va">SIG_BONF</span><span class="op">)</span> <span class="op">+</span>
                 <span class="fu">labs</span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Microscope: "</span>, <span class="va">MICROSCOPE</span>, <span class="st">"\nCovariates: None\nn samples: "</span>, <span class="va">N_SAMPLES</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span>

<span class="va">out_plot</span>
<span class="co">#&gt; Warning: Removed 7 rows containing missing values</span>
<span class="co">#&gt; (geom_point).</span></code></pre></div>
<div class="inline-figure"><img src="Period_intercept_with_AU_files/figure-html/unnamed-chunk-11-1.png" width="672"></div>
<div class="sourceCode" id="cb118"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggsave</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">PLOT_DIR</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">MICROSCOPE</span>, <span class="st">"_manhattan.png"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span>,
       <span class="va">out_plot</span>,
       device <span class="op">=</span> <span class="st">"png"</span>,
       width <span class="op">=</span> <span class="fl">9.6</span>,
       height <span class="op">=</span> <span class="fl">6</span>,
       units <span class="op">=</span> <span class="st">"in"</span>,
       dpi <span class="op">=</span> <span class="fl">400</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 7 rows containing missing values</span>
<span class="co">#&gt; (geom_point).</span></code></pre></div>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="gwas-results.html"><span class="header-section-number">8</span> GWAS results</a></div>
<div class="next"><a href="association-testing-with-only-db-microscope.html"><span class="header-section-number">10</span> Association testing with only DB microscope</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#association-testing-with-only-au-microscope"><span class="header-section-number">9</span> Association testing with only AU microscope</a></li>
<li><a class="nav-link" href="#load-libraries-and-variables"><span class="header-section-number">9.1</span> Load libraries and variables</a></li>
<li><a class="nav-link" href="#load-genotypes-and-positions"><span class="header-section-number">9.2</span> Load genotypes and positions</a></li>
<li><a class="nav-link" href="#read-in-phenotypes"><span class="header-section-number">9.3</span> Read in phenotypes</a></li>
<li><a class="nav-link" href="#filter-for-samples-with-both-genos-and-phenos"><span class="header-section-number">9.4</span> Filter for samples with both genos and phenos</a></li>
<li><a class="nav-link" href="#run-gwas"><span class="header-section-number">9.5</span> Run GWAS</a></li>
<li>
<a class="nav-link" href="#run-permutations"><span class="header-section-number">9.6</span> Run permutations</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#permute-phenos"><span class="header-section-number">9.6.1</span> Permute phenos</a></li>
<li><a class="nav-link" href="#run-gwls"><span class="header-section-number">9.6.2</span> Run GWLS</a></li>
<li><a class="nav-link" href="#get-minimum"><span class="header-section-number">9.6.3</span> Get minimum</a></li>
</ul>
</li>
<li><a class="nav-link" href="#generate-manhattan-plot"><span class="header-section-number">9.7</span> Generate Manhattan plot</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/brettellebi/somites/blob/master/Period_intercept_with_AU.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/brettellebi/somites/edit/master/Period_intercept_with_AU.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Somite period F2 analysis</strong>" was written by Ian Brettell. It was last built on 2022-03-14.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
