<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 8 Notes | Somite period F2 analysis</title>
<meta name="author" content="Ian Brettell">
<meta name="description" content="8.1 2 December 2021 Ali has sent the third batch of F2 data. Sample 323 is missing the first pair.  8.2 26 November 2021 Suggestions from EB: Set up guardrails – e.g. the site must have at least...">
<meta name="generator" content="bookdown 0.24.2 with bs4_book()">
<meta property="og:title" content="Chapter 8 Notes | Somite period F2 analysis">
<meta property="og:type" content="book">
<meta property="og:description" content="8.1 2 December 2021 Ali has sent the third batch of F2 data. Sample 323 is missing the first pair.  8.2 26 November 2021 Suggestions from EB: Set up guardrails – e.g. the site must have at least...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 8 Notes | Somite period F2 analysis">
<meta name="twitter:description" content="8.1 2 December 2021 Ali has sent the third batch of F2 data. Sample 323 is missing the first pair.  8.2 26 November 2021 Suggestions from EB: Set up guardrails – e.g. the site must have at least...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.10/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.5.1/tabs.js"></script><script src="libs/bs3compat-0.2.5.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="libs/plotly-binding-4.9.4.1/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.18/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Somite period F2 analysis</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="alignment-of-f0-and-f2-generations.html"><span class="header-section-number">2</span> Alignment of F0 and F2 generations</a></li>
<li><a class="" href="homozygosity-of-f0-cab-and-kaga.html"><span class="header-section-number">3</span> Homozygosity of F0 Cab and Kaga</a></li>
<li><a class="" href="f2-recombination-blocks-all-homozygous-divergent-sites.html"><span class="header-section-number">4</span> F2 recombination blocks (all homozygous-divergent sites)</a></li>
<li><a class="" href="f2-recombination-blocks-filtered-sites.html"><span class="header-section-number">5</span> F2 recombination blocks (filtered sites)</a></li>
<li><a class="" href="association-testing-with-simulated-phenotypes.html"><span class="header-section-number">6</span> Association testing with simulated phenotypes</a></li>
<li><a class="" href="gwls-results.html"><span class="header-section-number">7</span> GWLS results</a></li>
<li><a class="active" href="notes-1.html"><span class="header-section-number">8</span> Notes</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/brettellebi/somites">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="notes-1" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> Notes<a class="anchor" aria-label="anchor" href="#notes-1"><i class="fas fa-link"></i></a>
</h1>
<div id="december-2021" class="section level2" number="8.1">
<h2>
<span class="header-section-number">8.1</span> 2 December 2021<a class="anchor" aria-label="anchor" href="#december-2021"><i class="fas fa-link"></i></a>
</h2>
<p>Ali has sent the third batch of F2 data.</p>
<p>Sample 323 is missing the first pair.</p>
</div>
<div id="november-2021" class="section level2" number="8.2">
<h2>
<span class="header-section-number">8.2</span> 26 November 2021<a class="anchor" aria-label="anchor" href="#november-2021"><i class="fas fa-link"></i></a>
</h2>
<p>Suggestions from EB:</p>
<ul>
<li>Set up guardrails – e.g. the site must have at least 0.2 to 0.8 iCab vs Total</li>
<li>Limit to &gt;= 100 and &lt;= 10000</li>
<li>Plot chromosomes before filtering</li>
</ul>
</div>
<div id="november-2021-1" class="section level2" number="8.3">
<h2>
<span class="header-section-number">8.3</span> 17 November 2021<a class="anchor" aria-label="anchor" href="#november-2021-1"><i class="fas fa-link"></i></a>
</h2>
<p>Other HMM genotyping by sequencing (GBS) methods:</p>
<ul>
<li><p>Slides explaining HMM: <a href="https://alphagenes.roslin.ed.ac.uk/wp/wp-content/uploads/2019/05/05-JohnHickeyHMMIceCreamfastPHASEValenciaMay2018.pdf" class="uri">https://alphagenes.roslin.ed.ac.uk/wp/wp-content/uploads/2019/05/05-JohnHickeyHMMIceCreamfastPHASEValenciaMay2018.pdf</a></p></li>
<li>
<p>magicImpute</p>
<ul>
<li>Paper: <a href="https://academic.oup.com/genetics/article/210/1/71/6088020?login=true" class="uri">https://academic.oup.com/genetics/article/210/1/71/6088020?login=true</a>
</li>
<li>Implemented in RABBIT: <a href="https://github.com/chaozhi/RABBIT" class="uri">https://github.com/chaozhi/RABBIT</a>
</li>
</ul>
</li>
<li>
<p>R/qtl2</p>
<ul>
<li><a href="https://kbroman.org/qtl2/assets/vignettes/user_guide.htmlparalel" class="uri">https://kbroman.org/qtl2/assets/vignettes/user_guide.htmlparalel</a></li>
<li>Requires genotype file</li>
</ul>
</li>
<li>
<p>TIGER</p>
<ul>
<li>Paper: <a href="https://www.g3journal.org/content/5/3/385" class="uri">https://www.g3journal.org/content/5/3/385</a>
</li>
<li>GitHub: <a href="https://github.com/betharowan/TIGER_Scripts-for-distribution" class="uri">https://github.com/betharowan/TIGER_Scripts-for-distribution</a>
</li>
<li>Thesis describing TIGER: <a href="https://core.ac.uk/download/pdf/78377942.pdf" class="uri">https://core.ac.uk/download/pdf/78377942.pdf</a>
</li>
<li>Paper using TIGER for outbred founders: <a href="https://gsejournal.biomedcentral.com/articles/10.1186/s12711-019-0487-1" class="uri">https://gsejournal.biomedcentral.com/articles/10.1186/s12711-019-0487-1</a>
</li>
</ul>
</li>
<li>
<p>FSFHap</p>
<ul>
<li>Paper: <a href="https://acsess.onlinelibrary.wiley.com/doi/full/10.3835/plantgenome2014.05.0023" class="uri">https://acsess.onlinelibrary.wiley.com/doi/full/10.3835/plantgenome2014.05.0023</a>
</li>
<li>Implemented in TASSEL 5.0: <a href="https://tassel.bitbucket.io/" class="uri">https://tassel.bitbucket.io/</a>
</li>
<li>Appears specific to plants, as it infers parental sequence from the progency seqeunce bcause “the plants that were sequenced are usually not the same ones that were used to make the full sib population and may have somewhat different genotypes” (<a href="https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/FSFHapImputation/FSFHapImputation" class="uri">https://bitbucket.org/tasseladmin/tassel-5-source/wiki/UserManual/FSFHapImputation/FSFHapImputation</a>).</li>
<li>“Also, the progeny taken together generally provide very high coverage for the parental sequence and are often a better source for inferring the parent sequence than the parents themselves. FSFHap has a few different algorithms for inferring the parent haplotype. The default is the most robust and should be tried first for populations derived from F1’s.”</li>
<li>“The windowLD method works well for highly inbred parents, but has problems if one of the parents has residual heterozygosity. Additional options exist for backcross populations.”</li>
</ul>
</li>
<li>
<p>LB-Impute</p>
<ul>
<li>Imputes missing alleles with an HMM.</li>
<li>Paper: <a href="https://www.genetics.org/content/202/2/487" class="uri">https://www.genetics.org/content/202/2/487</a>
</li>
<li>GitHub: <a href="https://github.com/dellaporta-laboratory/LB-Impute" class="uri">https://github.com/dellaporta-laboratory/LB-Impute</a>
</li>
</ul>
</li>
<li>
<p>fastPHASE</p>
<ul>
<li>estimates missing genotypes and reconstructing haplotypes from unphased SNP genotype data of <em>unrelated</em> individuals.</li>
</ul>
</li>
</ul>
</div>
<div id="november-2021-2" class="section level2" number="8.4">
<h2>
<span class="header-section-number">8.4</span> 16 November 2021<a class="anchor" aria-label="anchor" href="#november-2021-2"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5345719/">https://www.ncbi.nlm.nih.gov/</a></li>
</ul>
<p><span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;&lt;strong&gt;rowan2015?&lt;/strong&gt;&lt;/p&gt;"><sup>2</sup></a></span> (TIGER)</p>
<ul>
<li><p>Trained Individual GenomE Reconstruction &gt; The number of crossovers
is typically limited to only one or two per chromosome pair per
meiosis.</p></li>
<li>
</ul>
<p><span class="citation"><a class="footnote-ref" tabindex="0" data-toggle="popover" data-content="&lt;p&gt;&lt;strong&gt;zan2019?&lt;/strong&gt;&lt;/p&gt;"><sup>3</sup></a></span></p>
<ul>
<li><p>Software / pipeline:
<a href="https://github.com/CarlborgGenomics/Stripes/tree/master/scripts" class="uri">https://github.com/CarlborgGenomics/Stripes/tree/master/scripts</a></p></li>
<li><p>Used 30x sequence data from founders, and &lt;0.5x sequence data on
intercross individuals to infer the founder mosaic genotypes of
intercross individuals.</p></li>
<li><p>7.6M markers segregated, and 10-13.7% of these fixed for alternative
alleles in the founders.</p></li>
<li><p>95% agreement between genotypes from low-coverage data, and those
obtained through SNP genotyping.</p></li>
<li><p>Estimated resolution of the inferred recombination breakpoints was
relatively high, with 50% of them defined on regions shorter than 10
kb.</p></li>
<li><p>Implements TIGER (referenced above)</p></li>
<li>
<li>
<p>Experimenting with calling F0 and F2 together, to get:</p>
<ul>
<li><p>Better calling accuracy for F0; and</p></li>
<li><p>To experiment with <em>bcftools</em>’s <code>roh</code> function
<a href="https://samtools.github.io/bcftools/bcftools.html#roh" class="uri">https://samtools.github.io/bcftools/bcftools.html#roh</a>, which
uses an HMM to call runs of homozygosity.</p></li>
<li>
<p>Also could test <em>plink1.9</em>’s <code>--homozyg</code> flag
<a href="https://www.cog-genomics.org/plink/1.9/ibd#homozyg" class="uri">https://www.cog-genomics.org/plink/1.9/ibd#homozyg</a></p>
<ul>
<li>With the algorithm description in <em>plink1.7</em>’s docs here:
<a href="https://zzz.bwh.harvard.edu/plink/ibdibs.shtml#homo" class="uri">https://zzz.bwh.harvard.edu/plink/ibdibs.shtml#homo</a>
</li>
</ul>
</li>
</ul>
</li>
<li>
<p>Differences in numbers of sites called:</p>
<ul>
<li>Original F0 only: 22618365 variants (18758908 SNPs)</li>
<li>F0 with F2:</li>
</ul>
</li>
<li><h2 id="bcftoolss-roh-results">
<em>bcftools</em>’s <code>roh</code> results:</h2></li>
</ul>
</div>
<div id="september-2021" class="section level2" number="8.5">
<h2>
<span class="header-section-number">8.5</span> 23 September 2021<a class="anchor" aria-label="anchor" href="#september-2021"><i class="fas fa-link"></i></a>
</h2>
<p>The following samples failed during the alignment of the second batch of
F2 individuals: <code>config/20210923_f2_fails.txt</code>. See that file for error
types.</p>
</div>
<div id="september-2021-1" class="section level2" number="8.6">
<h2>
<span class="header-section-number">8.6</span> 22 September 2021<a class="anchor" aria-label="anchor" href="#september-2021-1"><i class="fas fa-link"></i></a>
</h2>
<ul>
<li>Ali has finished uploading the second batch of F2 sequences.</li>
<li>Tom has given me the GridLMM code, which is saved to
<code>code/association_testing/gridLMM_gwas.R</code>
</li>
</ul>
</div>
<div id="august-2021" class="section level2" number="8.7">
<h2>
<span class="header-section-number">8.7</span> 31 August 2021<a class="anchor" aria-label="anchor" href="#august-2021"><i class="fas fa-link"></i></a>
</h2>
<div id="location-on-embl-server" class="section level3" number="8.7.1">
<h3>
<span class="header-section-number">8.7.1</span> Location on EMBL server:<a class="anchor" aria-label="anchor" href="#location-on-embl-server"><i class="fas fa-link"></i></a>
</h3>
<p>We use AnyConnect</p>
<p>vpn-gw1.embl.de</p>
<p>Server:</p>
<p><a href="smb://aulehla" class="uri">smb://aulehla</a></p>
<p>Username:</p>
<p>seleit</p>
<p>Location <a href="smb://aulehla/aulehla/NGS_Data_Library/2021-08-12-AAAG3YWHV" class="uri">smb://aulehla/aulehla/NGS_Data_Library/2021-08-12-AAAG3YWHV</a></p>
</div>
<div id="ascli" class="section level3" number="8.7.2">
<h3>
<span class="header-section-number">8.7.2</span> <code>ascli</code><a class="anchor" aria-label="anchor" href="#ascli"><i class="fas fa-link"></i></a>
</h3>
<p>They can use ascli from the terminal. All the info is at
<a href="https://intranet.embl.de/it_services/services/data_sharing/aspera/index.html" class="uri">https://intranet.embl.de/it_services/services/data_sharing/aspera/index.html</a>.</p>
<p><code>ascli</code> docs: <a href="https://www.rubydoc.info/gems/aspera-cli" class="uri">https://www.rubydoc.info/gems/aspera-cli</a></p>
<p>Examples from docs:</p>
<div class="sourceCode" id="cb81"><pre class="sourceCode ruby"><code class="sourceCode ruby"><span id="cb81-1"><a href="notes-1.html#cb81-1" aria-hidden="true" tabindex="-1"></a>ascli faspex health</span>
<span id="cb81-2"><a href="notes-1.html#cb81-2" aria-hidden="true" tabindex="-1"></a>ascli faspex package list</span>
<span id="cb81-3"><a href="notes-1.html#cb81-3" aria-hidden="true" tabindex="-1"></a>ascli faspex package list <span class="kw">--</span>box<span class="kw">=</span>sent <span class="kw">--</span>fields<span class="kw">=</span>package_id <span class="kw">--</span><span class="fu">format</span><span class="kw">=</span>csv <span class="kw">--</span>display<span class="kw">=</span>data<span class="kw">|</span>tail <span class="kw">-</span>n <span class="dv">1</span>);\</span>
<span id="cb81-4"><a href="notes-1.html#cb81-4" aria-hidden="true" tabindex="-1"></a>ascli faspex package list <span class="kw">--</span>fields<span class="kw">=</span>package_id <span class="kw">--</span><span class="fu">format</span><span class="kw">=</span>csv <span class="kw">--</span>display<span class="kw">=</span>data<span class="kw">|</span>tail <span class="kw">-</span>n <span class="dv">1</span>);\</span>
<span id="cb81-5"><a href="notes-1.html#cb81-5" aria-hidden="true" tabindex="-1"></a>ascli faspex package recv <span class="kw">--</span>to<span class="kw">-</span>folder<span class="kw">=.</span> <span class="kw">--</span>box<span class="kw">=</span>sent <span class="kw">--</span>id<span class="kw">=</span><span class="st">"my_package_id"</span></span>
<span id="cb81-6"><a href="notes-1.html#cb81-6" aria-hidden="true" tabindex="-1"></a>ascli faspex package recv <span class="kw">--</span>to<span class="kw">-</span>folder<span class="kw">=.</span> <span class="kw">--</span>id<span class="kw">=</span><span class="st">"my_package_id"</span></span>
<span id="cb81-7"><a href="notes-1.html#cb81-7" aria-hidden="true" tabindex="-1"></a>ascli faspex package recv <span class="kw">--</span>to<span class="kw">-</span>folder<span class="kw">=.</span> <span class="kw">--</span>id<span class="kw">=</span><span class="cn">ALL</span> <span class="kw">--</span>once<span class="kw">-</span>only<span class="kw">=</span>yes</span>
<span id="cb81-8"><a href="notes-1.html#cb81-8" aria-hidden="true" tabindex="-1"></a>ascli faspex package recv <span class="kw">--</span>to<span class="kw">-</span>folder<span class="kw">=.</span> <span class="kw">--</span>link<span class="kw">=</span><span class="st">"my_faspex_publink_recv_from_fxuser"</span></span>
<span id="cb81-9"><a href="notes-1.html#cb81-9" aria-hidden="true" tabindex="-1"></a>ascli faspex package send <span class="kw">--</span>delivery<span class="kw">-</span>info<span class="kw">=</span><span class="ot">@json</span>:<span class="vs">'{"title":"Important files delivery","recipients":["internal.user@example.com","FASPEX_USERNAME"]}'</span> testfile<span class="at">.bin</span></span>
<span id="cb81-10"><a href="notes-1.html#cb81-10" aria-hidden="true" tabindex="-1"></a>ascli faspex package send <span class="kw">--</span>link<span class="kw">=</span><span class="st">"my_faspex_publink_send_to_dropbox"</span> <span class="kw">--</span>delivery<span class="kw">-</span>info<span class="kw">=</span><span class="ot">@json</span>:<span class="vs">'{"title":"Important files delivery"}'</span> testfile<span class="at">.bin</span></span>
<span id="cb81-11"><a href="notes-1.html#cb81-11" aria-hidden="true" tabindex="-1"></a>ascli faspex package send <span class="kw">--</span>link<span class="kw">=</span><span class="st">"my_faspex_publink_send_to_fxuser"</span> <span class="kw">--</span>delivery<span class="kw">-</span>info<span class="kw">=</span><span class="ot">@json</span>:<span class="vs">'{"title":"Important files delivery"}'</span> testfile<span class="at">.bin</span></span>
<span id="cb81-12"><a href="notes-1.html#cb81-12" aria-hidden="true" tabindex="-1"></a>ascli faspex source name <span class="st">"Server Files"</span> node br <span class="ss">/</span></span>
<span id="cb81-13"><a href="notes-1.html#cb81-13" aria-hidden="true" tabindex="-1"></a><span class="ss">ascli faspex5 node list --value=@json:'{"type":"received","subtype":"mypackages"}'</span></span>
<span id="cb81-14"><a href="notes-1.html#cb81-14" aria-hidden="true" tabindex="-1"></a><span class="ss">ascli faspex5 package list --value=@json:'{"mailbox":"inbox","state":["released"]}'</span></span>
<span id="cb81-15"><a href="notes-1.html#cb81-15" aria-hidden="true" tabindex="-1"></a><span class="ss">ascli faspex5 package receive --id="my_package_id" --to-folder=.</span></span>
<span id="cb81-16"><a href="notes-1.html#cb81-16" aria-hidden="true" tabindex="-1"></a><span class="ss">ascli faspex5 package send --value=@json:'{"title":"test title","recipients":[{"name":"${f5_user}"}]}' testfile.bin</span></span></code></pre></div>
<p>Test:</p>
<div class="sourceCode" id="cb82"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb82-1"><a href="notes-1.html#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># On EBI cluster</span></span>
<span id="cb82-2"><a href="notes-1.html#cb82-2" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> /nfs/ftp/private/indigene_ftp/upload/Ali/Kaga-Cab_F2_Fish201-400_WGS/</span>
<span id="cb82-3"><a href="notes-1.html#cb82-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-4"><a href="notes-1.html#cb82-4" aria-hidden="true" tabindex="-1"></a><span class="co">## Example</span></span>
<span id="cb82-5"><a href="notes-1.html#cb82-5" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--link</span><span class="op">=</span><span class="st">"my_faspex_publink_recv_from_fxuser"</span></span>
<span id="cb82-6"><a href="notes-1.html#cb82-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-7"><a href="notes-1.html#cb82-7" aria-hidden="true" tabindex="-1"></a><span class="co">## Tests</span></span>
<span id="cb82-8"><a href="notes-1.html#cb82-8" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--link</span><span class="op">=</span><span class="st">"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo="</span></span>
<span id="cb82-9"><a href="notes-1.html#cb82-9" aria-hidden="true" tabindex="-1"></a><span class="co">#W, [2021-08-31T13:46:21.054223 #411589]  WARN -- : A new version is available: 4.2.0. You #have 4.1.0. Upgrade with: gem update </span></span>
<span id="cb82-10"><a href="notes-1.html#cb82-10" aria-hidden="true" tabindex="-1"></a><span class="co">#ascp: Error creating illegal char conversion table EINVAL, exiting.</span></span>
<span id="cb82-11"><a href="notes-1.html#cb82-11" aria-hidden="true" tabindex="-1"></a><span class="co">#E, [2021-08-31T13:46:27.329374 #411589] ERROR -- : code: 11</span></span>
<span id="cb82-12"><a href="notes-1.html#cb82-12" aria-hidden="true" tabindex="-1"></a><span class="co">#W, [2021-08-31T13:46:27.329485 #411589]  WARN -- : An error occured: ascp failed with code #1</span></span>
<span id="cb82-13"><a href="notes-1.html#cb82-13" aria-hidden="true" tabindex="-1"></a><span class="co">#W, [2021-08-31T13:46:27.329507 #411589]  WARN -- : resuming in  2 seconds (retry left:6)</span></span>
<span id="cb82-14"><a href="notes-1.html#cb82-14" aria-hidden="true" tabindex="-1"></a><span class="co">#ascp: Error creating illegal char conversion table EINVAL, exiting.</span></span>
<span id="cb82-15"><a href="notes-1.html#cb82-15" aria-hidden="true" tabindex="-1"></a><span class="co">#E, [2021-08-31T13:46:30.441640 #411589] ERROR -- : code: 11</span></span>
<span id="cb82-16"><a href="notes-1.html#cb82-16" aria-hidden="true" tabindex="-1"></a><span class="co">#W, [2021-08-31T13:46:30.441728 #411589]  WARN -- : An error occured: ascp failed with code #1</span></span>
<span id="cb82-17"><a href="notes-1.html#cb82-17" aria-hidden="true" tabindex="-1"></a><span class="co">#E, [2021-08-31T13:46:30.441748 #411589] ERROR -- : non-retryable error</span></span>
<span id="cb82-18"><a href="notes-1.html#cb82-18" aria-hidden="true" tabindex="-1"></a><span class="co">#ERROR: FASP(ascp]: ascp failed with code 1</span></span>
<span id="cb82-19"><a href="notes-1.html#cb82-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-20"><a href="notes-1.html#cb82-20" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--link</span><span class="op">=</span><span class="st">"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo=#"</span></span>
<span id="cb82-21"><a href="notes-1.html#cb82-21" aria-hidden="true" tabindex="-1"></a><span class="co"># As above</span></span>
<span id="cb82-22"><a href="notes-1.html#cb82-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-23"><a href="notes-1.html#cb82-23" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--link</span><span class="op">=</span><span class="st">"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo"</span></span>
<span id="cb82-24"><a href="notes-1.html#cb82-24" aria-hidden="true" tabindex="-1"></a><span class="co"># As above</span></span>
<span id="cb82-25"><a href="notes-1.html#cb82-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-26"><a href="notes-1.html#cb82-26" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--link</span><span class="op">=</span>https://faspex.embl.de/aspera/faspex/external_deliveries/4366<span class="pp">?</span>passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d<span class="kw">&amp;</span><span class="va">expiration=</span>MjAyMS0wOS0wN1QxMjo0Mjo0NVo=#</span>
<span id="cb82-27"><a href="notes-1.html#cb82-27" aria-hidden="true" tabindex="-1"></a><span class="co"># As above</span></span>
<span id="cb82-28"><a href="notes-1.html#cb82-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-29"><a href="notes-1.html#cb82-29" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--delivery-info</span><span class="op">=</span><span class="st">"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo="</span></span>
<span id="cb82-30"><a href="notes-1.html#cb82-30" aria-hidden="true" tabindex="-1"></a><span class="co"># ERROR: Argument: Missing mandatory option: id</span></span>
<span id="cb82-31"><a href="notes-1.html#cb82-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-32"><a href="notes-1.html#cb82-32" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--id</span><span class="op">=</span>brettell@ebi.ac.uk <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--delivery-info</span><span class="op">=</span><span class="st">"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo="</span></span>
<span id="cb82-33"><a href="notes-1.html#cb82-33" aria-hidden="true" tabindex="-1"></a><span class="co"># ERROR: Argument: Missing mandatory option: url</span></span>
<span id="cb82-34"><a href="notes-1.html#cb82-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-35"><a href="notes-1.html#cb82-35" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--id</span><span class="op">=</span>brettell@ebi.ac.uk <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--url</span><span class="op">=</span><span class="st">"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo="</span></span>
<span id="cb82-36"><a href="notes-1.html#cb82-36" aria-hidden="true" tabindex="-1"></a><span class="co"># ERROR: Argument: Missing mandatory option: username</span></span>
<span id="cb82-37"><a href="notes-1.html#cb82-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-38"><a href="notes-1.html#cb82-38" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--id</span><span class="op">=</span>brettell@ebi.ac.uk <span class="at">--username</span><span class="op">=</span>brettell@ebi.ac.uk <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--url</span><span class="op">=</span><span class="st">"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo="</span></span>
<span id="cb82-39"><a href="notes-1.html#cb82-39" aria-hidden="true" tabindex="-1"></a><span class="co"># ERROR: Argument: Missing mandatory option: password</span></span>
<span id="cb82-40"><a href="notes-1.html#cb82-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-41"><a href="notes-1.html#cb82-41" aria-hidden="true" tabindex="-1"></a><span class="ex">ascli</span> faspex package recv <span class="at">--to-folder</span><span class="op">=</span>. <span class="at">--transfer</span><span class="op">=</span>httpgw <span class="at">--transfer-info</span><span class="op">=</span>@json:<span class="st">'{"url":"https://faspex.embl.de/aspera/faspex/external_deliveries/4366?passcode=221f3482a4d9b2fd87da2d21a2916156c2c1870d&amp;expiration=MjAyMS0wOS0wN1QxMjo0Mjo0NVo=#"}'</span></span></code></pre></div>
</div>
<div id="lftp" class="section level3" number="8.7.3">
<h3>
<span class="header-section-number">8.7.3</span> <code>lftp</code><a class="anchor" aria-label="anchor" href="#lftp"><i class="fas fa-link"></i></a>
</h3>
<p>Hi, You can use the lftp command present in datatransfer.embl.de (here
is a tutorial)
<a href="https://linuxconfig.org/lftp-tutorial-on-linux-with-examples" class="uri">https://linuxconfig.org/lftp-tutorial-on-linux-with-examples</a> . If you
want to send files that are in your group share, they will be available
at /g/ location. Cheers.</p>
<hr>
<p>Josep Manel Andrés Moscardó Systems Engineer, IT Operations EMBL
Heidelberg T +49 6221 387-8394</p>
</div>
</div>
<div id="august-2021-1" class="section level2" number="8.8">
<h2>
<span class="header-section-number">8.8</span> 3 August 2021<a class="anchor" aria-label="anchor" href="#august-2021-1"><i class="fas fa-link"></i></a>
</h2>
<p>Added Cab and Kaga karyoplots.</p>
<p>Removed use of (almost) all wrappers in Snakemake pipeline. Now all
rules are containerised other than for pulling the reference sequence
from Ensembl.</p>
<p><span color="red">NOTE</span>: Install <code>plotly</code> and <code>DT</code> in R container.</p>
</div>
<div id="july-2021" class="section level2" number="8.9">
<h2>
<span class="header-section-number">8.9</span> 15 July 2021<a class="anchor" aria-label="anchor" href="#july-2021"><i class="fas fa-link"></i></a>
</h2>
<p>Notes for Ali:</p>
<ul>
<li>Tried sample 171 again (the one you uploaded again), and it failed
again at the alignment stage with the following error:</li>
</ul>
<div class="sourceCode" id="cb83"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb83-1"><a href="notes-1.html#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="ex">[mem_sam_pe_batch_post]</span> paired reads have different names: <span class="st">"�0�����O&amp;�����x�7�Z�HĻ�@�%Q�ü��"</span>, <span class="st">"ST-K00119:220:HKNVLBBXY:7:1101:1661:1138"</span></span></code></pre></div>
<ul>
<li><p>Fixed the sample names so that they don’t include the pool number at
the beginning, e.g. sample 171 is no longer called 7171</p></li>
<li><p>Cab has 15M homozygous sites, whereas Kaga only has 5M. Reason: Cab
aligns better to HdrR, so there are many more 0/0 calls?</p></li>
<li><p>3.48M homozygous sites after joining Cab and Kaga homozygous site
DFs.</p></li>
<li><p>2.7M homozygous divergent sites.</p></li>
</ul>
</div>
<div id="july-2021-1" class="section level2" number="8.10">
<h2>
<span class="header-section-number">8.10</span> 13 July 2021<a class="anchor" aria-label="anchor" href="#july-2021-1"><i class="fas fa-link"></i></a>
</h2>
<div id="attendees" class="section level3" number="8.10.1">
<h3>
<span class="header-section-number">8.10.1</span> Attendees:<a class="anchor" aria-label="anchor" href="#attendees"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>Ewan Birney</li>
<li>Alexander Aulehla</li>
<li>Tomas Fitzgerald</li>
<li>Ali Seleit</li>
<li>Ian Brettell</li>
</ul>
</div>
<div id="notes-2" class="section level3" number="8.10.2">
<h3>
<span class="header-section-number">8.10.2</span> Notes:<a class="anchor" aria-label="anchor" href="#notes-2"><i class="fas fa-link"></i></a>
</h3>
<ul>
<li>
<p>Genetics:</p>
<ul>
<li>
<p>Increase block size?</p>
<ul>
<li>Try site-by-site</li>
<li>HMM would need to work on a <em>count</em> system so that it
doesn’t get skewed</li>
</ul>
</li>
</ul>
</li>
<li>
<p><span color="red"><strong>ACTION</strong></span> (Ian):</p>
<ul>
<li>
<p>Check colours:</p>
<ul>
<li>Would expect Cab skew, especially as it has the reporter.</li>
</ul>
</li>
</ul>
</li>
<li>
<p><span color="red"><strong>ACTION</strong></span> (Ian):</p>
<ul>
<li><p>Find number of reads covering each block for each individual.</p></li>
<li><p>If they reach &lt; 5, increase the block size.</p></li>
<li><p>Also check out chr 16 reporter locus.</p></li>
<li><p>Make histogram of counts per block for each individual.</p></li>
<li>
<p>Data frame:</p>
<ul>
<li>Fish ID</li>
<li>Block ID</li>
<li>Allele</li>
<li>Count</li>
</ul>
</li>
<li><p>Plot a boxplot of counts per allele for all blocks per
individual.</p></li>
<li><p>Do for 5, 10 and 15 kb windows, then make boxplots to compare.</p></li>
<li><p>Expect coverage across the homo-divergent regions is quite
uneven.</p></li>
<li><p>Also estimate heritability?</p></li>
</ul>
</li>
<li>
<p><span color="red"><strong>ACTION</strong></span> (Ali):</p>
<ul>
<li><p>Eyeball high phenotype points</p></li>
<li><p>Also look at the plate position.</p></li>
<li><p>Is there a between-day effect? (Note: Can only image 10
individuals per day)</p></li>
<li>
<p>Run a Kruskal-Wallis test for:</p>
<ul>
<li>day</li>
<li>microscope (if they differ)</li>
<li>any other way you can split it up</li>
<li>look at the within-individual variance too</li>
</ul>
</li>
</ul>
</li>
<li>
<p>Re: sample size:</p>
<ul>
<li>Pause imaging while we check the above.</li>
<li>Want to see the parents tight, and the F2 stretching out.</li>
</ul>
</li>
</ul>
</div>
</div>
</div>

  <div class="chapter-nav">
<div class="prev"><a href="gwls-results.html"><span class="header-section-number">7</span> GWLS results</a></div>
<div class="next"><a href="references.html">References</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#notes-1"><span class="header-section-number">8</span> Notes</a></li>
<li><a class="nav-link" href="#december-2021"><span class="header-section-number">8.1</span> 2 December 2021</a></li>
<li><a class="nav-link" href="#november-2021"><span class="header-section-number">8.2</span> 26 November 2021</a></li>
<li><a class="nav-link" href="#november-2021-1"><span class="header-section-number">8.3</span> 17 November 2021</a></li>
<li><a class="nav-link" href="#november-2021-2"><span class="header-section-number">8.4</span> 16 November 2021</a></li>
<li><a class="nav-link" href="#september-2021"><span class="header-section-number">8.5</span> 23 September 2021</a></li>
<li><a class="nav-link" href="#september-2021-1"><span class="header-section-number">8.6</span> 22 September 2021</a></li>
<li>
<a class="nav-link" href="#august-2021"><span class="header-section-number">8.7</span> 31 August 2021</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#location-on-embl-server"><span class="header-section-number">8.7.1</span> Location on EMBL server:</a></li>
<li><a class="nav-link" href="#ascli"><span class="header-section-number">8.7.2</span> ascli</a></li>
<li><a class="nav-link" href="#lftp"><span class="header-section-number">8.7.3</span> lftp</a></li>
</ul>
</li>
<li><a class="nav-link" href="#august-2021-1"><span class="header-section-number">8.8</span> 3 August 2021</a></li>
<li><a class="nav-link" href="#july-2021"><span class="header-section-number">8.9</span> 15 July 2021</a></li>
<li>
<a class="nav-link" href="#july-2021-1"><span class="header-section-number">8.10</span> 13 July 2021</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#attendees"><span class="header-section-number">8.10.1</span> Attendees:</a></li>
<li><a class="nav-link" href="#notes-2"><span class="header-section-number">8.10.2</span> Notes:</a></li>
</ul>
</li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/brettellebi/somites/blob/master/Notes.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/brettellebi/somites/edit/master/Notes.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Somite period F2 analysis</strong>" was written by Ian Brettell. It was last built on 2021-12-17.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
