<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 8 16 November 2021 | Meeting log</title>
<meta name="author" content="Ian Brettell">
<meta name="description" content="https://www.ncbi.nlm.nih.gov/ Yanjun Zan et al.2 Other methods: * ‘Genotyping by low-coverage whole-genome sequencing in intercross pedigrees from outbred founders: a cost-efficient approach’:...">
<meta name="generator" content="bookdown 0.24.2 with bs4_book()">
<meta property="og:title" content="Chapter 8 16 November 2021 | Meeting log">
<meta property="og:type" content="book">
<meta property="og:description" content="https://www.ncbi.nlm.nih.gov/ Yanjun Zan et al.2 Other methods: * ‘Genotyping by low-coverage whole-genome sequencing in intercross pedigrees from outbred founders: a cost-efficient approach’:...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 8 16 November 2021 | Meeting log">
<meta name="twitter:description" content="https://www.ncbi.nlm.nih.gov/ Yanjun Zan et al.2 Other methods: * ‘Genotyping by low-coverage whole-genome sequencing in intercross pedigrees from outbred founders: a cost-efficient approach’:...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/header-attrs-2.10/header-attrs.js"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.5.1/tabs.js"></script><script src="libs/bs3compat-0.2.5.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.3/htmlwidgets.js"></script><script src="libs/plotly-binding-4.9.4.1/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/crosstalk-1.1.1/css/crosstalk.css" rel="stylesheet">
<script src="libs/crosstalk-1.1.1/js/crosstalk.min.js"></script><link href="libs/plotly-htmlwidgets-css-1.57.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-1.57.1/plotly-latest.min.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.18/datatables.js"></script><link href="libs/dt-core-1.10.20/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.10.20/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.10.20/js/jquery.dataTables.min.js"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Meeting log</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled"><li><a class="" href="index.html"><span class="header-section-number">Chapter 8</span> 16 November 2021</a></li></ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/brettellebi/somites">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="november-2021" class="section level1" number="8">
<h1>
<span class="header-section-number">8</span> 16 November 2021<a class="anchor" aria-label="anchor" href="#november-2021"><i class="fas fa-link"></i></a>
</h1>
<ul>
<li><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5345719/">https://www.ncbi.nlm.nih.gov/</a></li>
<li><span class="citation">Yanjun Zan et al.<a class="footnote-ref" tabindex="0" data-toggle="popover" data-content='&lt;p&gt;&lt;span&gt;“Genotyping by Low-Coverage Whole-Genome Sequencing in Intercross Pedigrees from Outbred Founders: A Cost-Efficient Approach,”&lt;/span&gt; &lt;em&gt;Genetics Selection Evolution&lt;/em&gt; 51, no. 1 (August 14, 2019): 44, &lt;a href="https://doi.org/10.1186/s12711-019-0487-1" role="doc-biblioref"&gt;https://doi.org/10.1186/s12711-019-0487-1&lt;/a&gt;.&lt;/p&gt;'><sup>2</sup></a></span></li>
</ul>
<p>Other methods: * ‘Genotyping by low-coverage whole-genome sequencing in
intercross pedigrees from outbred founders: a cost-efficient approach’:
<a href="https://gsejournal.biomedcentral.com/articles/10.1186/s12711-019-0487-1" class="uri">https://gsejournal.biomedcentral.com/articles/10.1186/s12711-019-0487-1</a>
-</p>
<ul>
<li>
<p>Experimenting with calling F0 and F2 together, to get:</p>
<ul>
<li><p>Better calling accuracy for F0; and</p></li>
<li><p>To experiment with <em>bcftools</em>’s <code>roh</code> function
<a href="https://samtools.github.io/bcftools/bcftools.html#roh" class="uri">https://samtools.github.io/bcftools/bcftools.html#roh</a>, which
uses an HMM to call runs of homozygosity.</p></li>
<li>
<p>Also could test <em>plink1.9</em>’s <code>--homozyg</code> flag
<a href="https://www.cog-genomics.org/plink/1.9/ibd#homozyg" class="uri">https://www.cog-genomics.org/plink/1.9/ibd#homozyg</a></p>
<ul>
<li>With the algorithm description in <em>plink1.7</em>’s docs here:
<a href="https://zzz.bwh.harvard.edu/plink/ibdibs.shtml#homo" class="uri">https://zzz.bwh.harvard.edu/plink/ibdibs.shtml#homo</a>
</li>
</ul>
</li>
</ul>
</li>
<li>
<p>Differences in numbers of sites called:</p>
<ul>
<li>Original F0 only: 22618365 variants (18758908 SNPs)</li>
<li>F0 with F2:</li>
</ul>
</li>
<li><h2 id="bcftoolss-roh-results">
<em>bcftools</em>’s <code>roh</code> results:</h2></li>
</ul>
</div>

  <div class="chapter-nav">
<div class="empty"></div>
<div class="empty"></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <div id="book-on-this-page"></div>

      <div class="book-extra">
        <ul class="list-unstyled">
          
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Meeting log</strong>" was written by Ian Brettell. It was last built on 2021-11-16.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
