<!DOCTYPE html>
<html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Chapter 12 Reporter as covariate | Somite period F2 analysis</title>
<meta name="author" content="Ian Brettell">
<meta name="description" content="Run GWAS on split inverse-normalised microscope, with reporter genotype as a covariate.  12.1 Setup library(tidyverse) library(GridLMM) library(KRLS) GENO_FILE =...">
<meta name="generator" content="bookdown 0.25 with bs4_book()">
<meta property="og:title" content="Chapter 12 Reporter as covariate | Somite period F2 analysis">
<meta property="og:type" content="book">
<meta property="og:description" content="Run GWAS on split inverse-normalised microscope, with reporter genotype as a covariate.  12.1 Setup library(tidyverse) library(GridLMM) library(KRLS) GENO_FILE =...">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Chapter 12 Reporter as covariate | Somite period F2 analysis">
<meta name="twitter:description" content="Run GWAS on split inverse-normalised microscope, with reporter genotype as a covariate.  12.1 Setup library(tidyverse) library(GridLMM) library(KRLS) GENO_FILE =...">
<!-- JS --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://kit.fontawesome.com/6ecbd6c532.js" crossorigin="anonymous"></script><script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="libs/bootstrap-4.6.0/bootstrap.min.css" rel="stylesheet">
<script src="libs/bootstrap-4.6.0/bootstrap.bundle.min.js"></script><script src="libs/bs3compat-0.2.5.1/tabs.js"></script><script src="libs/bs3compat-0.2.5.1/bs3compat.js"></script><link href="libs/bs4_book-1.0.0/bs4_book.css" rel="stylesheet">
<script src="libs/bs4_book-1.0.0/bs4_book.js"></script><script src="libs/htmlwidgets-1.5.4/htmlwidgets.js"></script><script src="libs/plotly-binding-4.10.0/plotly.js"></script><script src="libs/typedarray-0.1/typedarray.min.js"></script><link href="libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
<script src="libs/crosstalk-1.2.0/js/crosstalk.min.js"></script><link href="libs/plotly-htmlwidgets-css-2.5.1/plotly-htmlwidgets.css" rel="stylesheet">
<script src="libs/plotly-main-2.5.1/plotly-latest.min.js"></script><link href="libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
<script src="libs/datatables-binding-0.21/datatables.js"></script><link href="libs/dt-core-1.11.3/css/jquery.dataTables.min.css" rel="stylesheet">
<link href="libs/dt-core-1.11.3/css/jquery.dataTables.extra.css" rel="stylesheet">
<script src="libs/dt-core-1.11.3/js/jquery.dataTables.min.js"></script><style>
    #content {
        font-size: 0.9rem;
    }
    </style>
<script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- CSS --><style type="text/css">
    
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  </style>
<style type="text/css">
    /* Used with Pandoc 2.11+ new --citeproc when CSL is used */
    div.csl-bib-body { }
    div.csl-entry {
      clear: both;
        }
    .hanging div.csl-entry {
      margin-left:2em;
      text-indent:-2em;
    }
    div.csl-left-margin {
      min-width:2em;
      float:left;
    }
    div.csl-right-inline {
      margin-left:2em;
      padding-left:1em;
    }
    div.csl-indent {
      margin-left: 2em;
    }
  </style>
<link rel="stylesheet" href="style.css">
</head>
<body data-spy="scroll" data-target="#toc">

<div class="container-fluid">
<div class="row">
  <header class="col-sm-12 col-lg-3 sidebar sidebar-book"><a class="sr-only sr-only-focusable" href="#content">Skip to main content</a>

    <div class="d-flex align-items-start justify-content-between">
      <h1>
        <a href="index.html" title="">Somite period F2 analysis</a>
      </h1>
      <button class="btn btn-outline-primary d-lg-none ml-2 mt-1" type="button" data-toggle="collapse" data-target="#main-nav" aria-expanded="true" aria-controls="main-nav"><i class="fas fa-bars"></i><span class="sr-only">Show table of contents</span></button>
    </div>

    <div id="main-nav" class="collapse-lg">
      <form role="search">
        <input id="search" class="form-control" type="search" placeholder="Search" aria-label="Search">
</form>

      <nav aria-label="Table of contents"><h2>Table of contents</h2>
        <ul class="book-toc list-unstyled">
<li><a class="" href="index.html"><span class="header-section-number">1</span> About</a></li>
<li><a class="" href="alignment-of-f0-and-f2-generations.html"><span class="header-section-number">2</span> Alignment of F0 and F2 generations</a></li>
<li><a class="" href="homozygosity-of-f0-cab-and-kaga.html"><span class="header-section-number">3</span> Homozygosity of F0 Cab and Kaga</a></li>
<li><a class="" href="f2-recombination-blocks-all-homozygous-divergent-sites.html"><span class="header-section-number">4</span> F2 recombination blocks (all homozygous-divergent sites)</a></li>
<li><a class="" href="f2-recombination-blocks-filtered-sites.html"><span class="header-section-number">5</span> F2 recombination blocks (filtered sites)</a></li>
<li><a class="" href="phenotypes.html"><span class="header-section-number">6</span> Phenotypes</a></li>
<li><a class="" href="association-testing-with-simulated-phenotypes.html"><span class="header-section-number">7</span> Association testing with simulated phenotypes</a></li>
<li><a class="" href="gwas-results.html"><span class="header-section-number">8</span> GWAS results</a></li>
<li><a class="" href="association-testing-with-only-au-microscope.html"><span class="header-section-number">9</span> Association testing with only AU microscope</a></li>
<li><a class="" href="association-testing-with-only-db-microscope.html"><span class="header-section-number">10</span> Association testing with only DB microscope</a></li>
<li><a class="" href="inverse-normalize-by-microscope.html"><span class="header-section-number">11</span> Inverse-normalize by microscope</a></li>
<li><a class="active" href="reporter-as-covariate.html"><span class="header-section-number">12</span> Reporter as covariate</a></li>
<li><a class="" href="annotations-of-gwas-hits.html"><span class="header-section-number">13</span> Annotations of GWAS hits</a></li>
<li><a class="" href="references.html">References</a></li>
</ul>

        <div class="book-extra">
          <p><a id="book-repo" href="https://github.com/brettellebi/somites">View book source <i class="fab fa-github"></i></a></p>
        </div>
      </nav>
</div>
  </header><main class="col-sm-12 col-md-9 col-lg-7" id="content"><div id="reporter-as-covariate" class="section level1" number="12">
<h1>
<span class="header-section-number">12</span> Reporter as covariate<a class="anchor" aria-label="anchor" href="#reporter-as-covariate"><i class="fas fa-link"></i></a>
</h1>
<p>Run GWAS on split inverse-normalised microscope, with reporter genotype as a covariate.</p>
<div id="setup" class="section level2" number="12.1">
<h2>
<span class="header-section-number">12.1</span> Setup<a class="anchor" aria-label="anchor" href="#setup"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org">tidyverse</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">GridLMM</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://www.r-project.org">KRLS</a></span><span class="op">)</span>

<span class="va">GENO_FILE</span> <span class="op">=</span> <span class="st">"/hps/nobackup/birney/users/ian/somites/association_testing/20220321/all_sites/inputs/5000.rds"</span>
<span class="va">PHENO_FILE</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"data/20220321_phenotypes.xlsx"</span><span class="op">)</span> <span class="co"># True phenotypes</span>
<span class="va">GWLS_SOURCE_FILE</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"workflow/scripts/run_gwas_source.R"</span><span class="op">)</span>
<span class="va">MANHAT_SOURCE_FILE</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"workflow/scripts/get_manhattan_source.R"</span><span class="op">)</span>
<span class="va">BIN_LENGTH</span> <span class="op">=</span> <span class="fl">5000</span>
<span class="va">TARGET_PHENO</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"intercept"</span><span class="op">)</span>
<span class="va">MICROSCOPE</span> <span class="op">=</span> <span class="st">"split_invnorm_reporter"</span>
<span class="va">OUT_PHENOS</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"data/20220310_invnorm_phenos.csv"</span><span class="op">)</span>
<span class="va">PLOT_DIR</span> <span class="op">=</span> <span class="fu">here</span><span class="fu">::</span><span class="fu"><a href="https://here.r-lib.org//reference/here.html">here</a></span><span class="op">(</span><span class="st">"book/plots/20220321/microscope_test"</span><span class="op">)</span>
<span class="va">ALPHA</span> <span class="op">=</span> <span class="fl">0.05</span>
<span class="va">REPORTER_LOC</span> <span class="op">=</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>CHROM <span class="op">=</span> <span class="fl">16</span>,
                              START <span class="op">=</span> <span class="fl">28706898</span>,
                              END <span class="op">=</span> <span class="fl">28708417</span><span class="op">)</span>

<span class="co"># Get GWAS functions</span>

<span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="va">GWLS_SOURCE_FILE</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="va">MANHAT_SOURCE_FILE</span><span class="op">)</span>
<span class="co"># Get lighter/darker functions</span>
<span class="kw"><a href="https://rdrr.io/r/base/source.html">source</a></span><span class="op">(</span><span class="st">"https://gist.githubusercontent.com/brettellebi/c5015ee666cdf8d9f7e25fa3c8063c99/raw/91e601f82da6c614b4983d8afc4ef399fa58ed4b/karyoploteR_lighter_darker.R"</span><span class="op">)</span></code></pre></div>
</div>
<div id="read-in-data-2" class="section level2" number="12.2">
<h2>
<span class="header-section-number">12.2</span> Read in data<a class="anchor" aria-label="anchor" href="#read-in-data-2"><i class="fas fa-link"></i></a>
</h2>
<div id="genos" class="section level3" number="12.2.1">
<h3>
<span class="header-section-number">12.2.1</span> Genos<a class="anchor" aria-label="anchor" href="#genos"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb145"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Load genotypes and positions</span>
<span class="va">in_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="va">GENO_FILE</span><span class="op">)</span></code></pre></div>
</div>
<div id="phenos" class="section level3" number="12.2.2">
<h3>
<span class="header-section-number">12.2.2</span> Phenos<a class="anchor" aria-label="anchor" href="#phenos"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Read in phenotypes</span>

<span class="co">## Read in file and wrangle</span>
<span class="va">phenos</span> <span class="op">=</span> <span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html">read_xlsx</a></span><span class="op">(</span><span class="va">PHENO_FILE</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># adjust sample names</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>SAMPLE <span class="op">=</span> <span class="va">fish</span> <span class="op">%&gt;%</span> <span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_remove.html">str_remove</a></span><span class="op">(</span><span class="st">"KC"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
                    <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># ensure that the phenotype column is numeric</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu">all_of</span><span class="op">(</span><span class="va">TARGET_PHENO</span><span class="op">)</span>,
                                <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html">as.numeric</a></span><span class="op">(</span><span class="va">.x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Split into two DFs by microscope and inverse-normalise</span>
<span class="va">phenos_au</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Microscope</span> <span class="op">==</span> <span class="st">"AU"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="va">TARGET_PHENO</span><span class="op">)</span>,
                              <span class="op">~</span><span class="fu">my.invnorm</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">phenos_db</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Microscope</span> <span class="op">==</span> <span class="st">"DB"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="va">TARGET_PHENO</span><span class="op">)</span>,
                              <span class="op">~</span><span class="fu">my.invnorm</span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Combine</span>
<span class="va">phenos</span> <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span><span class="va">phenos_au</span>,
                          <span class="va">phenos_db</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">SAMPLE</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>SAMPLE <span class="op">=</span> <span class="va">SAMPLE</span> <span class="op">%&gt;%</span> 
                  <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Write to file</span>
<span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html">write_csv</a></span><span class="op">(</span><span class="va">phenos</span>, <span class="va">OUT_PHENOS</span><span class="op">)</span>

<span class="co"># Plot</span>
<span class="va">micr_pal</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"#8D99AE"</span>, <span class="st">"#2b2d42"</span><span class="op">)</span>

<span class="va">pheno_hist</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span> 
  <span class="co"># remove NAs in `Microscope` column</span>
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">Microscope</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_histogram</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">intercept</span>, fill <span class="op">=</span> <span class="va">Microscope</span><span class="op">)</span>,
                   bins <span class="op">=</span> <span class="fl">50</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="va">micr_pal</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">facet_grid</span><span class="op">(</span>rows <span class="op">=</span> <span class="fu">vars</span><span class="op">(</span><span class="va">Microscope</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"period intercept"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Inverse-normalised period intercept split by microscope"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">guides</span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span>

<span class="va">pheno_hist</span></code></pre></div>
<div class="inline-figure"><img src="Reporter_as_covariate_files/figure-html/unnamed-chunk-3-1.png" width="672"></div>
</div>
<div id="covariates" class="section level3" number="12.2.3">
<h3>
<span class="header-section-number">12.2.3</span> Covariates<a class="anchor" aria-label="anchor" href="#covariates"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb147"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># How many counts for each reporter genotype? </span>
<span class="va">phenos</span> <span class="op">%&gt;%</span>
  <span class="fu">count</span><span class="op">(</span><span class="va">reporter_geno</span><span class="op">)</span>
<span class="co">#&gt; # A tibble: 4 × 2</span>
<span class="co">#&gt;   reporter_geno     n</span>
<span class="co">#&gt;           &lt;dbl&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1            -1   341</span>
<span class="co">#&gt; 2             0   249</span>
<span class="co">#&gt; 3             1     1</span>
<span class="co">#&gt; 4            NA    47</span></code></pre></div>
<div class="sourceCode" id="cb148"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot</span>
<span class="va">phenos</span> <span class="op">%&gt;%</span> 
  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>reporter_geno <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="va">reporter_geno</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">reporter_geno</span>, <span class="va">intercept</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">geom_violin</span><span class="op">(</span>fill <span class="op">=</span> <span class="va">intercept_pal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span> 
    <span class="fu">geom_boxplot</span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.3</span>, fill <span class="op">=</span> <span class="va">intercept_pal</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggbeeswarm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_beeswarm.html">geom_beeswarm</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.4</span>, alpha <span class="op">=</span> <span class="fl">0.5</span>, colour <span class="op">=</span> <span class="fu">lighter</span><span class="op">(</span><span class="va">intercept_pal</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">xlab</span><span class="op">(</span><span class="st">"Reporter genotype"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ylab</span><span class="op">(</span><span class="st">"Inverse-normalised intercept split by microscope"</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu">ggtitle</span><span class="op">(</span><span class="st">"Intercept phenotype by reporter genotype"</span><span class="op">)</span>
<span class="co">#&gt; Warning: Groups with fewer than two data points have been</span>
<span class="co">#&gt; dropped.</span></code></pre></div>
<div class="inline-figure"><img src="Reporter_as_covariate_files/figure-html/unnamed-chunk-5-1.png" width="672"></div>
</div>
<div id="prepare-input-1" class="section level3" number="12.2.4">
<h3>
<span class="header-section-number">12.2.4</span> Prepare input<a class="anchor" aria-label="anchor" href="#prepare-input-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb149"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">## Filter and order phenotypes</span>
<span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span>
    <span class="co"># filter phenotypes for those with genotypes</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">SAMPLE</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># join to `sample_order` to ensure phenotypes are in the correct order   </span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>SAMPLE <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                     <span class="va">.</span>,
                     by <span class="op">=</span> <span class="st">"SAMPLE"</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># remove NAs (created by the samples that have genotypes but not phenotypes)</span>
    <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html">drop_na</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="co"># the GridLMM code doesn't work with tibbles</span>
    <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>

<span class="co">## Filter genotypes for those that have phenotypes</span>
<span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span>
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span> <span class="op">%&gt;%</span> 
                   <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>

<span class="co">## Filter sample_order for those that have phenotypes</span>
<span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span>

<span class="co">## Get number of samples</span>
<span class="va">N_SAMPLES</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-1" class="section level3" number="12.2.5">
<h3>
<span class="header-section-number">12.2.5</span> Run<a class="anchor" aria-label="anchor" href="#run-1"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb150"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Run GWAS</span>
<span class="va">out</span> <span class="op">=</span> <span class="fu">run_gwas</span><span class="op">(</span>d <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span>,
               m <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"positions"</span><span class="op">]</span><span class="op">]</span>,
               p <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span>,
               covariates <span class="op">=</span> <span class="st">"reporter_geno"</span>
              <span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">out</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/gwas_results/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb151"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/gwas_results/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Rename p-value column</span>
<span class="kw">if</span> <span class="op">(</span><span class="st">"p_value_REML"</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span><span class="op">(</span><span class="va">out</span><span class="op">$</span><span class="va">results</span><span class="op">)</span><span class="op">)</span><span class="op">{</span>
  <span class="va">P_COL</span> <span class="op">=</span> <span class="st">"p_value_REML"</span>
<span class="op">}</span> <span class="kw">else</span> <span class="op">{</span>
  <span class="va">P_COL</span> <span class="op">=</span> <span class="st">"p_value_REML.1"</span>
<span class="op">}</span>

<span class="co"># Rename column in results</span>

<span class="kw">if</span> <span class="op">(</span><span class="va">P_COL</span> <span class="op">==</span> <span class="st">"p_value_REML.1"</span><span class="op">)</span><span class="op">{</span>
  <span class="va">out</span><span class="op">$</span><span class="va">results</span> <span class="op">=</span> <span class="va">out</span><span class="op">$</span><span class="va">results</span> <span class="op">%&gt;%</span> 
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html">rename</a></span><span class="op">(</span>p_value_REML <span class="op">=</span> <span class="va">p_value_REML.1</span><span class="op">)</span>
<span class="op">}</span></code></pre></div>
</div>
</div>
<div id="run-permutations-3" class="section level2" number="12.3">
<h2>
<span class="header-section-number">12.3</span> Run permutations<a class="anchor" aria-label="anchor" href="#run-permutations-3"><i class="fas fa-link"></i></a>
</h2>
<div id="permute-phenos-3" class="section level3" number="12.3.1">
<h3>
<span class="header-section-number">12.3.1</span> Permute phenos<a class="anchor" aria-label="anchor" href="#permute-phenos-3"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb152"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">seeds</span> <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>

<span class="va">counter</span> <span class="op">=</span> <span class="fl">0</span>
<span class="va">perm_phenos</span> <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">seeds</span>, <span class="kw">function</span><span class="op">(</span><span class="va">SEED</span><span class="op">)</span><span class="op">{</span>
  <span class="va">counter</span> <span class="op">&lt;&lt;-</span> <span class="va">counter</span> <span class="op">+</span> <span class="fl">1</span>
  <span class="co"># set seed</span>
  <span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="va">seeds</span><span class="op">[</span><span class="va">counter</span><span class="op">]</span><span class="op">)</span>
  <span class="co"># randomise</span>
  <span class="co"># Randomise all columns other than `fish`</span>
  <span class="va">permute_order</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">phenos</span><span class="op">)</span><span class="op">)</span>
  <span class="va">out</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span> 
    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/across.html">across</a></span><span class="op">(</span><span class="op">-</span><span class="va">SAMPLE</span>,
                                <span class="op">~</span><span class="va">.x</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/order.html">order</a></span><span class="op">(</span><span class="va">permute_order</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
                  <span class="op">)</span>
<span class="op">}</span><span class="op">)</span></code></pre></div>
</div>
<div id="run-gwls-3" class="section level3" number="12.3.2">
<h3>
<span class="header-section-number">12.3.2</span> Run GWLS<a class="anchor" aria-label="anchor" href="#run-gwls-3"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb153"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm_out</span> <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map</a></span><span class="op">(</span><span class="va">perm_phenos</span>, <span class="kw">function</span><span class="op">(</span><span class="va">PERM_PHENO</span><span class="op">)</span><span class="op">{</span>
  <span class="co">## Read in file and wrangle</span>
  <span class="va">phenos</span> <span class="op">=</span> <span class="va">PERM_PHENO</span>
  
  <span class="co">## Filter and order phenotypes</span>
  <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">phenos</span> <span class="op">%&gt;%</span>
      <span class="co"># filter phenotypes for those with genotypes</span>
      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">SAMPLE</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="co"># join to `sample_order` to ensure phenotypes are in the correct order   </span>
      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span><span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>SAMPLE <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,
                       <span class="va">.</span>,
                       by <span class="op">=</span> <span class="st">"SAMPLE"</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="co"># remove NAs (created by the samples that have genotypes but not phenotypes)</span>
      <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/drop_na.html">drop_na</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="co"># the GridLMM code doesn't work with tibbles</span>
      <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">as.data.frame</a></span><span class="op">(</span><span class="op">)</span>
  
  <span class="co">## Filter genotypes for those that have phenotypes</span>
  <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span> <span class="op">%&gt;%</span>
      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html">slice</a></span><span class="op">(</span><span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">%in%</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span> <span class="op">%&gt;%</span> 
                     <span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>
  
  <span class="co">## Filter sample_order for those that have phenotypes</span>
  <span class="va">in_list</span><span class="op">[[</span><span class="st">"sample_order"</span><span class="op">]</span><span class="op">]</span> <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">SAMPLE</span>
              
  <span class="co"># Run GWAS</span>
  
  <span class="va">out</span> <span class="op">=</span> <span class="fu">run_gwas</span><span class="op">(</span>d <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"genotypes"</span><span class="op">]</span><span class="op">]</span>,
                 m <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"positions"</span><span class="op">]</span><span class="op">]</span>,
                 p <span class="op">=</span> <span class="va">in_list</span><span class="op">[[</span><span class="st">"phenotypes"</span><span class="op">]</span><span class="op">]</span>,
                 covariates <span class="op">=</span> <span class="st">"reporter_geno"</span>
                <span class="op">)</span>
  
  <span class="kw"><a href="https://rdrr.io/r/base/function.html">return</a></span><span class="op">(</span><span class="va">out</span><span class="op">)</span>
<span class="op">}</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">perm_out</span><span class="op">)</span> <span class="op">=</span> <span class="va">seeds</span>

<span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">saveRDS</a></span><span class="op">(</span><span class="va">perm_out</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/perms/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb154"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm_out</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"/hps/nobackup/birney/users/ian/somites/microscope_test/perms/"</span>, <span class="va">MICROSCOPE</span>, <span class="st">".rds"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="get-minimum-3" class="section level3" number="12.3.3">
<h3>
<span class="header-section-number">12.3.3</span> Get minimum<a class="anchor" aria-label="anchor" href="#get-minimum-3"><i class="fas fa-link"></i></a>
</h3>
<div class="sourceCode" id="cb155"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">perm_df</span> <span class="op">=</span> <span class="fu">purrr</span><span class="fu">::</span><span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html">map_dfr</a></span><span class="op">(</span><span class="va">perm_out</span>, <span class="kw">function</span><span class="op">(</span><span class="va">PERM</span><span class="op">)</span><span class="op">{</span>
  <span class="va">OUT</span> <span class="op">=</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html">tibble</a></span><span class="op">(</span>MIN_P <span class="op">=</span> <span class="va">PERM</span><span class="op">$</span><span class="va">results</span> <span class="op">%&gt;%</span> 
                         <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/all_of.html">all_of</a></span><span class="op">(</span><span class="va">P_COL</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
                         <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">.</span>, na.rm <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>
  <span class="op">)</span>
<span class="op">}</span>, .id <span class="op">=</span> <span class="st">"SEED"</span><span class="op">)</span>

<span class="co"># Get minimum</span>
<span class="va">SIG_LEVEL</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span><span class="op">(</span><span class="va">perm_df</span><span class="op">$</span><span class="va">MIN_P</span><span class="op">)</span>

<span class="co"># Get bonferroni correction</span>
<span class="va">SIG_BONF</span> <span class="op">=</span> <span class="va">ALPHA</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">out</span><span class="op">$</span><span class="va">results</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="generate-manhattan-plot-3" class="section level2" number="12.4">
<h2>
<span class="header-section-number">12.4</span> Generate Manhattan plot<a class="anchor" aria-label="anchor" href="#generate-manhattan-plot-3"><i class="fas fa-link"></i></a>
</h2>
<div class="sourceCode" id="cb156"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">out_clean</span> <span class="op">=</span> <span class="fu">clean_gwas_res</span><span class="op">(</span><span class="va">out</span>,
                           bin_length <span class="op">=</span> <span class="va">BIN_LENGTH</span>,
                           chr_lens <span class="op">=</span> <span class="va">med_chr_lens</span><span class="op">)</span>


<span class="co"># Get bonferroni significance level</span>
<span class="va">BONFERRONI</span> <span class="op">=</span> <span class="va">ALPHA</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">nrow</a></span><span class="op">(</span><span class="va">out_clean</span><span class="op">)</span>

<span class="co"># Plot</span>
<span class="va">out_plot</span> <span class="op">=</span> <span class="fu">plot_man</span><span class="op">(</span><span class="va">out_clean</span>,
                    site_filter <span class="op">=</span> <span class="st">"all_sites"</span>,
                    phenotype <span class="op">=</span> <span class="va">TARGET_PHENO</span>,
                    bin_length <span class="op">=</span> <span class="va">BIN_LENGTH</span>, 
                    gwas_pal <span class="op">=</span> <span class="va">intercept_pal</span>,
                    med_chr_lens <span class="op">=</span> <span class="va">med_chr_lens</span>,
                    sig_level <span class="op">=</span> <span class="va">SIG_LEVEL</span>,
                    bonferroni <span class="op">=</span> <span class="va">BONFERRONI</span><span class="op">)</span> <span class="op">+</span>
                 <span class="co">#ylim(0,7) + </span>
                 <span class="fu">labs</span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="st">"Microscope: split_invnorm"</span>, <span class="st">"\nCovariates: Reporter genotype\nn samples: "</span>, <span class="va">N_SAMPLES</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span>

<span class="va">out_plot</span>
<span class="co">#&gt; Warning: Removed 9 rows containing missing values</span>
<span class="co">#&gt; (geom_point).</span></code></pre></div>
<div class="inline-figure"><img src="Reporter_as_covariate_files/figure-html/unnamed-chunk-13-1.png" width="672"></div>
<div class="sourceCode" id="cb157"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">ggsave</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">PLOT_DIR</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste</a></span><span class="op">(</span><span class="va">MICROSCOPE</span>, <span class="st">"_manhattan.png"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span>,
       <span class="va">out_plot</span>,
       device <span class="op">=</span> <span class="st">"png"</span>,
       width <span class="op">=</span> <span class="fl">9.6</span>,
       height <span class="op">=</span> <span class="fl">6</span>,
       units <span class="op">=</span> <span class="st">"in"</span>,
       dpi <span class="op">=</span> <span class="fl">400</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 9 rows containing missing values</span>
<span class="co">#&gt; (geom_point).</span></code></pre></div>

</div>
</div>
  <div class="chapter-nav">
<div class="prev"><a href="inverse-normalize-by-microscope.html"><span class="header-section-number">11</span> Inverse-normalize by microscope</a></div>
<div class="next"><a href="annotations-of-gwas-hits.html"><span class="header-section-number">13</span> Annotations of GWAS hits</a></div>
</div></main><div class="col-md-3 col-lg-2 d-none d-md-block sidebar sidebar-chapter">
    <nav id="toc" data-toggle="toc" aria-label="On this page"><h2>On this page</h2>
      <ul class="nav navbar-nav">
<li><a class="nav-link" href="#reporter-as-covariate"><span class="header-section-number">12</span> Reporter as covariate</a></li>
<li><a class="nav-link" href="#setup"><span class="header-section-number">12.1</span> Setup</a></li>
<li>
<a class="nav-link" href="#read-in-data-2"><span class="header-section-number">12.2</span> Read in data</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#genos"><span class="header-section-number">12.2.1</span> Genos</a></li>
<li><a class="nav-link" href="#phenos"><span class="header-section-number">12.2.2</span> Phenos</a></li>
<li><a class="nav-link" href="#covariates"><span class="header-section-number">12.2.3</span> Covariates</a></li>
<li><a class="nav-link" href="#prepare-input-1"><span class="header-section-number">12.2.4</span> Prepare input</a></li>
<li><a class="nav-link" href="#run-1"><span class="header-section-number">12.2.5</span> Run</a></li>
</ul>
</li>
<li>
<a class="nav-link" href="#run-permutations-3"><span class="header-section-number">12.3</span> Run permutations</a><ul class="nav navbar-nav">
<li><a class="nav-link" href="#permute-phenos-3"><span class="header-section-number">12.3.1</span> Permute phenos</a></li>
<li><a class="nav-link" href="#run-gwls-3"><span class="header-section-number">12.3.2</span> Run GWLS</a></li>
<li><a class="nav-link" href="#get-minimum-3"><span class="header-section-number">12.3.3</span> Get minimum</a></li>
</ul>
</li>
<li><a class="nav-link" href="#generate-manhattan-plot-3"><span class="header-section-number">12.4</span> Generate Manhattan plot</a></li>
</ul>

      <div class="book-extra">
        <ul class="list-unstyled">
<li><a id="book-source" href="https://github.com/brettellebi/somites/blob/master/book/Reporter_as_covariate.Rmd">View source <i class="fab fa-github"></i></a></li>
          <li><a id="book-edit" href="https://github.com/brettellebi/somites/edit/master/book/Reporter_as_covariate.Rmd">Edit this page <i class="fab fa-github"></i></a></li>
        </ul>
</div>
    </nav>
</div>

</div>
</div> <!-- .container -->

<footer class="bg-primary text-light mt-5"><div class="container"><div class="row">

  <div class="col-12 col-md-6 mt-3">
    <p>"<strong>Somite period F2 analysis</strong>" was written by Ian Brettell. It was last built on 2022-03-26.</p>
  </div>

  <div class="col-12 col-md-6 mt-3">
    <p>This book was built by the <a class="text-light" href="https://bookdown.org">bookdown</a> R package.</p>
  </div>

</div></div>
</footer>
</body>
</html>
