# Exploration of hits

```{r, message = F, warning = F}
library(tidyverse)
library(ggbeeswarm)
```


## Period intercept

### Read in files

```{r}
HITS = "/hps/nobackup/birney/users/ian/somites/gcta/mlma_loco/true/hdrr/None/5000/0.8/unsegmented_psm_area/None.loco.mlma"
GENOS = "/hps/nobackup/birney/users/ian/somites/hmm_out/F2/hdrr/hmmlearn_true/None/5000/0.8.csv"
PHENOS = "/hps/nobackup/birney/users/ian/somites/phens/true/unsegmented_psm_area.phen"
BIN_LENGTH = 5000

# Read in hits and pull location of top hit
hits = readr::read_tsv(HITS) %>% 
  dplyr::rename(BIN = bp) %>% 
  dplyr::mutate(BIN_START = (BIN * BIN_LENGTH) + 1,
                BIN_END = ((BIN + 1) * BIN_LENGTH)) %>% 
  # make beta direction consistent
  dplyr::mutate(ADJ_BETA = dplyr::case_when(A1 == "B" ~ b,
                                            A1 == "A" ~ -b))

top_hit = hits %>% 
  dplyr::slice(which.min(p)) 

TARGET_CHROM = top_hit %>% 
  dplyr::pull(Chr)

TARGET_BIN = top_hit %>% 
  dplyr::pull(BIN)

genos = readr::read_csv(GENOS) %>% 
  dplyr::filter(CHROM == TARGET_CHROM & BIN == TARGET_BIN)

geno_means = readr::read_csv(GENOS) %>% 
  dplyr::group_by(SAMPLE) %>% 
  dplyr::summarise(MEAN_PROP_KAGA = mean(PROP_KAGA))

phenos = readr::read_tsv(PHENOS,
                         col_names = c("FID", "SAMPLE", "PHENO"))

# Join

df = dplyr::left_join(genos %>% 
                        dplyr::select(SAMPLE, STATE),
                      phenos %>% 
                        dplyr::select(SAMPLE, PHENO),
                      by = "SAMPLE")

df_prop = dplyr::left_join(geno_means,
                           phenos %>% 
                             dplyr::select(SAMPLE, PHENO),
                           by = "SAMPLE")
  
```

### Plot

#### Mean proportion of Kaga against phenotype

```{r}
df_prop %>% 
  ggplot(aes(MEAN_PROP_KAGA, PHENO)) +
    geom_point() +
    geom_smooth() +
    theme_bw() +
  xlab("Mean proportion of Kaga reads across all bins") +
  ylab("unsegmented PSM area")
```


#### Top hit

```{r}
df %>% 
  dplyr::mutate(STATE = factor(STATE, levels = 0:2)) %>% 
  ggplot(aes(STATE, PHENO)) +
    geom_boxplot() +
    ggbeeswarm::geom_beeswarm(alpha = 0.5) +
    theme_bw() +
    ggtitle(paste("PSM area top hit",
                  "\nLocus: ",
                  TARGET_CHROM,
                  ":", 
                  (TARGET_BIN * BIN_LENGTH + 1),
                  "-",
                  ((TARGET_BIN + 1) * BIN_LENGTH),
                  sep = "")) +
    xlab("Genotype (0 = Hom Cab, 1 = Het, 2 = Hom Kaga)") +
    ylab("PSM area")
```

#### All betas

```{r}
hits %>% 
  dplyr::mutate(Chr = factor(Chr, levels = 1:24)) %>% 
  ggplot() +
  geom_point(aes(ADJ_BETA, -log10(p), colour = Chr), alpha = 0.1) +
  theme_bw() +
  facet_wrap(~Chr, ncol = 4) +
  xlab("beta for Kaga allele") +
  ylab(expression(-log[10](p))) +
  ggtitle("Phenotype: PSM area") +
  guides(colour = "none")
```

```{r}
# Calculate percentages
pos_neg_perc = hits %>% 
  dplyr::mutate(POS_NEG = dplyr::case_when(ADJ_BETA > 0 ~ "pos",
                                           ADJ_BETA < 0 ~ "neg")) %>% 
  dplyr::count(POS_NEG) %>% 
  dplyr::add_count(wt = n, name = "nn") %>% 
  dplyr::mutate(PERC = n / nn * 100)

TITLE = "Phenotype: PSM area"
SUBTITLE = paste("Percent positive: ",
              pos_neg_perc %>% 
                dplyr::filter(POS_NEG == "pos") %>% 
                dplyr::pull(PERC),
              "\nPercent negative: ",
              pos_neg_perc %>% 
                dplyr::filter(POS_NEG == "neg") %>% 
                dplyr::pull(PERC),
              sep = "")

hits %>% 
  ggplot() +
    geom_density(aes(ADJ_BETA)) +
    theme_bw() +
    xlab("beta for Kaga allele") +
    ggtitle(TITLE,
            subtitle = SUBTITLE) 
```

